### ISSUE TITLE
Feature Request - Rate limiting based on time of day

### ISSUE THREAD

**Issue by Raynok** [ncw/rclone#221]
Hi There,
Hope this is something that others would be interested in.

Looking for the ability to control data transfer rates based on time of day. 

Best Regards
Rayno

**Comment by ncw** [comment]
Hi @Raynok  - you wrote me an email about this didnt' you which got lost in my inbox - sorry! Good choice making a ticket ;-)

What do you think about a [command line interface a bit like this](https://code.google.com/p/bwlimit/wiki/bwlimit)?

**Comment by Raynok** [comment]
Hi Nick,
No worries, Alex and Klaus pointed me in the right direction. 
I think, "Absolutely fantastic".

Going to give it a try and will let you. Great Project !!

**Comment by Raynok** [comment]
Hi Nick,
Thank you for the pointer, there is a issue when issuing the command. Windows based computers.

My original command 
rclone.exe copy Software:GLOBALSOFTWARE c:\support\ --transfers=1 -c --bwlimit=100
error: None

Testing command
rclone.exe copy Software:GLOBALSOFTWARE c:\support\ --transfers=1 -c --bwlimit=100 7-21,50"
Error: It is only necessary to use a unique prefix of the subcommand, eg 'mo'for 'move'.
Command copy needs 2 arguments maximum

If its something simple - I do apologize . I've tried use a .bat file with a set variable and got the same error.

**Comment by ncw** [comment]
That syntax `-bwlimit=100 7-21,50` won't work at the moment, I just wondered what you thought about that syntax?

Namely having a 1 hour granularity and setting it like that.

If we think it is a good idea then we can implement something

**Comment by Raynok** [comment]
The syntax makes sense and pretty simple to grasp though there must be some reference in the documentation ;-) . Setting the 1 hour granularity is perfect - I would do the same. You have my vote :+1:

**Comment by utdrmac** [comment]
+1 on this as well. That way while I'm sleeping, rclone can use up all my bandwidth without me having to mess with it.

**Comment by AlexisEvo** [comment]
I whipped up some systemd timers to do this, as well as to make sure folders are automatically backed up.

https://gist.github.com/AlexisEvo/3a803de79332b4e5fb83

It's gross, but it works. :)

**Comment by Raynok** [comment]
Howzit - Still using rlcone and enjoying the new versions.

Alexis :+1: nice one . Didn't understand at first but did some reading. My environments are 99.9% Microsoft.

 I've created a deployment package to distribute rclone to specific computers.

1.) The deployed package creates the folders and shares. Then...
2.) Downloads rclone + config file from google drive with a download rate I've set.
3.) Initiates the download of the files and logs it to a file.
4.) Once the download is complete it uploads the log file with the computer name and timestamp.

This gives me the ability to see the progress across multi pal wan sites and if there are no log files then its either busy downloading still or failed. 

In a nut shell.

Control all the files on google and it mimics at the destinations. I can see the activity logs in google and this allows me to manage and view multi pal wan sites under one plane of glass (not perfect but best effort)

Thought I share how i try use rclone.

**Comment by arionl** [comment]
Not sure if it's helpful to anyone, but I'm using some rudimentary traffic shaping in Linux to help with bandwidth management while this native capability in rclone is being matured. Here's a link to some info: http://lartc.org/howto/lartc.ratelimit.single.html

Instead of the last line in the example that shapes based on destination IP I'm just shaping based on destination port (443) using this command:

```tc filter add dev $DEV parent 1: protocol ip prio 16 u32 match ip dport 443 0xffff flowid 1:1```

The Linux box I'm using for ```rclone sync``` commands is dedicated so I'm not too worried about shaping other port 443 traffic. It's hard to shape based on destination IP because most of these cloud services have a multitude of endpoints..

**Comment by marcopaganini** [comment]
Why not use a multiple hh:mm,bandwidth type of configuration, having the currently selected bandwidth limits stay until superseded by the next configuration time slot? Something like:

-bwtimes="23:00,0 05:00,3000 08:00,500 20:00,3000"

This means:
- At 5am, switch to 3000
- At 8am, switch to 500
- At 8pm, switch to 3000
- At 11pm, switch to unlimited

Order should be irrelevant.

This format should be easy to parse and provides substantial flexibility with time selection.

**Comment by ncw** [comment]
@marcopaganini Does that mean you set `--bwlimit` as well as `--bwtimes`? I quite like there being only one command line, so as you suggested, but with the default bwlimit first, so something like `--bwlmit ="500k 23:00,off 05:00,3M 08:00,500k 20:00,3M"` (the `SizeSuffix` parser will parse the `3M` or `off` for you).

**Comment by marcopaganini** [comment]
That works well, but what is the purpose of the "default" bwlimit in this
case? If we have time slots, we'll always be covered by one of them at a
given time of the day.

**Comment by ncw** [comment]
@marcopaganini I guess the default limit is what you get until you get to one of the trigger times when the bandwidth limit changes. If you don't have that then you'll have to work out what the current limit should be at any given time.  You'll probably have to work that out anyway though - I imagine a go routine which runs every minute which resets the bwlimit to the limit appropriate to that time.

I think having two flags for setting the bwlimit may confuse people, but I'm happy to go with your suggestion for the exact syntax.

**Comment by marcopaganini** [comment]
I totally agree with having one flag. I'm just a bit confused about the
"default" bwlimit. The way I was envisioning this was to always scan the
"time slots" and see where we currently fit, setting the bw limiter to that
value.

So, say, someone says 08:00,1M, 23:00,5M, it means that at any point in the
day, we'll be doing 1M (08:00 to 22:59) or 5M (23:00 to 07:59). Of course,
if only one time slot is used, then we'll be at that limit for the whole
day. :)

**Comment by ncw** [comment]
@marcopaganini I mis-understood your original plan - I thought it was an ordered list of instructions of time:bwlimit, but I see now that it is a timetable and yes I agree with you now :-)

**Comment by marcopaganini** [comment]
Ooops, my bad! I should have made it more clear :) But yeah, the idea is
that it's a timetable. At any point, we're inside one of its "slots".

**Comment by ncw** [comment]
@marcopaganini are you still working on this?

**Comment by marcopaganini** [comment]
Hello Nick, haven't touched it yet (lack of time). I'll see if I find some
time to work on this during the holidays.

**Comment by marcopaganini** [comment]
So, FYI I have a change coded, but it's still not working yet. PR coming
soon.

**Comment by ncw** [comment]
Here is a beta with this new feature in: http://beta.rclone.org/v1.35-10-g3b0f944/ (uploaded in 15-30 mins)

Here are the new docs

### --bwlimit=BANDWIDTH_SPEC ###

This option controls the bandwidth limit. Limits can be specified
in two ways: As a single limit, or as a timetable.

Single limits last for the duration of the session. To use a single limit,
specify the desired bandwidth in kBytes/s, or use a suffix b|k|M|G.  The
default is `0` which means to not limit bandwidth.

For example to limit bandwidth usage to 10 MBytes/s use `--bwlimit 10M`

It is also possible to specify a "timetable" of limits, which will cause
certain limits to be applied at certain times. To specify a timetable, format your
entries as "HH:MM,BANDWIDTH HH:MM,BANDWITH...".

An example of a typical timetable to avoid link saturation during daytime
working hours could be:

`--bwlimit "08:00,512 12:00,10M 13:00,512 18:00,30M 23:00,off"`

In this example, the transfer bandwidth will be set to 512kBytes/sec at 8am.
At noon, it will raise to 10Mbytes/s, and drop back to 512kBytes/sec at 1pm.
At 6pm, the bandwidth limit will be set to 30MBytes/s, and at 11pm it will be
completely disabled (full speed). Anything between 11pm and 8am will remain
unlimited.

Bandwidth limits only apply to the data transfer. The don't apply to the
bandwith of the directory listings etc.

Note that the units are Bytes/s not Bits/s.  Typically connections are
measured in Bits/s - to convert divide by 8.  For example let's say
you have a 10 Mbit/s connection and you wish rclone to use half of it
- 5 Mbit/s.  This is 5/8 = 0.625MByte/s so you would use a `--bwlimit
0.625M` parameter for rclone.

**Comment by friendlytherapist** [comment]
This is only in beta right?

**Comment by ncw** [comment]
@friendlytherapist it will be in the 1.36 release, currently in the beta, yes.

**Comment by calisro** [comment]
I'm having a bit of a problem with shell escaping.  

I use a variable to store a bunch of my default switches but no matter how I escape the quotes, don't use them, or use different quotes, it isn't interpretting like i'd expect.

Example:
 STD_SWITCHES="--stats 5m \
   --config='$RCLONECONF' \
   --transfers=16 --checkers=32 \
   --one-file-system \
   --bwlimit=\"07:00,6M 23:00,10M\" "

-># echo $STD_SWITCHES
--stats 5m --config='/home/robert/.rclone.conf' --transfers=16 --checkers=32 --one-file-system --bwlimit="07:00,6M 23:00,10M"
root@HS: ~
-># rclone $STD_SWITCHES   rmdirs 
Error: invalid argument "\"07:00,6M" for --bwlimit="07:00,6M: invalid time specification (hh:mm): "\"07:00"
Usage:
...
...
2017/02/27 12:14:21 Fatal error: invalid argument "\"07:00,6M" for --bwlimit="07:00,6M: invalid time specification (hh:mm): "\"07:00"


I've tried all of the following:
STD_SWITCHES="--stats 5m \
   --config='$RCLONECONF' \
   --transfers=16 --checkers=32 \
   --one-file-system \
   --bwlimit='07:00,6M 23:00,10M' "

STD_SWITCHES="--stats 5m \
   --config='$RCLONECONF' \
   --transfers=16 --checkers=32 \
   --one-file-system \
   --bwlimit=07:00,6M 23:00,10M "

STD_SWITCHES="--stats 5m \
   --config='$RCLONECONF' \
   --transfers=16 --checkers=32 \
   --one-file-system \
   --bwlimit=07:00,6M\ 23:00,10M "

**Comment by calisro** [comment]
I opened a new issue for it instead  Shell escaping #1183

**Comment by Raynok** [comment]
Thank you for this feature, appreciate it.
